ceph/ceph_admin/helper.py

```diff
@@ -2,10 +2,183 @@
 """
 import logging
+from os.path import dirname
+
+from jinja2 import Template
+
+from ceph.utils import get_node_by_id
 
 LOG = logging.getLogger()
 
 
+class UnknownSpecFound(Exception):
+    pass
+
+
+class PrepareSpec:
+    """prepare spec file based on provided config"""
+
+    COMMON_SERVICES = [
+        "mon",
+        "mgr",
+        "alertmanager",
+        "crash",
+        "grafana",
+        "node-exporter",
+        "prometheus",
+    ]
+
+    def __init__(self, cluster, specs):
+        """
+        Initialize the PrepareSpec
+
+        Args:
+            cluster: ceph cluster (ceph-nodes)
+            specs: service specifications
+
+        specs:
+          - service_type: host
+            address: true
+            labels: apply-all-labels
+            nodes:
+                - node2
+                - node3
+          - service_type: mon
+            placement:
+              nodes:
+                - node2
+                - node3
+          - service_type: mgr
+            placement:
+                count: 2
+          - service_type: alertmanager
+            placement:
+                count: 1
+          - service_type: crash
+            placement:
+                host_pattern: '*'
+          - service_type: grafana
+            placement:
+                count: 1
+          - service_type: node-exporter
+            placement:
+                host_pattern: '*'
+          - service_type: prometheus
+            placement:
+                count: 1
+        """
+        self.cluster = cluster
+        self.specs = specs
+        self.template_path = dirname(__file__) + "/jinja_templates/"
+
+    def _get_template(self, service_type):
+        """
+        Return Jinja template based on the service_type
+        Args:
+            service_type: service name (ex., "host")
+        Returns:
+            template
+        """
+        path = self.template_path + f"{service_type}.jinja"
+        with open(path) as fd:
+            template = fd.read()
+        return Template(template)
+
+    def render_hosts(self, spec):
+        """
+        prepare hosts spec content based on host config
+        args:
+            spec: hosts specification
+
+        spec:
+          - service_type: host
+            address: true
+            labels: apply-all-labels
+            nodes:
+                - node2
+                - node3
+        Returns:
+            hosts_spec
+        """
+        template = self._get_template("host")
+        hosts = []
+        address = spec.get("address")
+        labels = spec.get("labels")
+        for node_name in spec["nodes"]:
+            host = dict()
+            node = get_node_by_id(self.cluster, node_name)
+            host["hostname"] = node.shortname
+            if address:
+                host["address"] = node.ip_address
+            if labels:
+                host["labels"] = node.role.role_list
+            hosts.append(host)
+
+        return template.render(hosts=hosts)
```

Comment:
This can be moved to `class Host: apply_spec`

Reply:
As we discussed, apply_spec is part of Orch.